libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2248,8 +2248,17 @@ void TypeChecker::typeCheckABIEncodeCallFunction(FunctionCall const& _functionCa
 	auto const* tupleType = dynamic_cast<TupleType const*>(type(*arguments[1]));
 	if (tupleType)
 	{
-		auto const& argumentTuple = dynamic_cast<TupleExpression const&>(*arguments[1].get());
-		callArguments = decltype(callArguments){argumentTuple.components().begin(), argumentTuple.components().end()};
+		if (TupleExpression const* argumentTuple = dynamic_cast<TupleExpression const*>(arguments[1].get()))
+			callArguments = decltype(callArguments){argumentTuple->components().begin(), argumentTuple->components().end()};
+		else
+		{
+			m_errorReporter.typeError(
+				9062_error,
+				arguments[1]->location(),
+				"This type cannot be encoded."
```

Comment:
The message is identical to `2056_error` and in that case I'd expect both to have the same error code.

But actually, the checks are not identical. Maybe we should just make the message here more specific?
```suggestion
				"Expected an inline tuple, not an expression of a tuple type."
```

Or maybe this. Might be clearer to users but is a bit informal. Not sure which one's better:
```suggestion
				"Expected an inline tuple embedded directly in the call, not an expression of a tuple type."
```

Reply:
Right. I was not sure about the message so I just left it as in the other similar error.
About your suggestions, I think that the first one is good. The origin of the error will be highlighted by the compiler message, so I guess it should be obvious that the inline tuple is expected to be embedded inside the call:

```
Error: Expected an inline tuple, not an expression of a tuple type.
  --> /path/to/file.sol:x:y
   |
13 |         abi.encodeCall(this.f0, g0());
   |                                 ^^^^
```